File: AlgorithmStudy/15주차/BOJ20922/김재환20922.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        int[] arr = new int[N];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        int[] cnt = new int[100001];
        int left = 0, right = 0;
        int ans = 0;

        while (right < N) {
            if (cnt[arr[right]] < K) {
                cnt[arr[right]]++;
                right++;
                ans = Math.max(ans, right - left);
            } else {
                cnt[arr[left]]--;
                left++;
            }
        }
        System.out.println(ans);
    }
}

File: AlgorithmStudy/15주차/BOJ20922/김재환20922.py
import sys
input = sys.stdin.readline

N, K = map(int, input().split())
arr = list(map(int, input().split()))

cnt = [0] * 100001
left = 0
right = 0
answer = 0

while right < N:
    # 현재 right에 있는 숫자 개수를 추가해도 K개 이하라면
    if cnt[arr[right]] < K:
        cnt[arr[right]] += 1
        right += 1
        answer = max(answer, right - left)
    else:
        # 넘친다면 left를 하나씩 줄인다
        cnt[arr[left]] -= 1
        left += 1

print(answer)
